
## Opinion Structure in Sentiment Analysis

Opinions in sentiment analysis are often broken down into components such as **entity**, **aspect**, **sentiment**, **opinion holder**, and **time**. Using a product review example:

**Review Example:**
Posted by: *Sarah Lee*  
Date: *March 5, 2022*  
“(1) I bought the Galaxy ZX Pro smartphone three months ago.  
(2) I absolutely love using it.  
(3) Its display is crystal clear.  
(4) The battery easily lasts all day.  
(5) However, my brother says it feels too bulky.”

### Key Observations
- **Multiple Opinions:** Positive and negative sentiments about the same product.
- **Two People:** Sarah and her brother express opinions.
- **Time Stamp:** Review posted on March 5, 2022.

### Opinion Analysis
- Sentence (1): Objective statement (Neutral).
- Sentence (2): Positive opinion about the smartphone overall.
- Sentence (3): Positive sentiment about the display.
- Sentence (4): Positive sentiment about battery performance.
- Sentence (5): Negative opinion about size (bulkiness), attributed to another person.

### Formal Representation
An opinion can be represented as: Opinion = (g, s) <br>

Where:
- **g** = Target (entity or aspect, e.g., product, feature)
- **s** = Sentiment orientation (positive, negative, neutral) or numerical rating.

**Example:**
- Sentence (2): Target = *Galaxy ZX Pro*, Sentiment = *Positive*.


### Quintable notation:

Opinion = (eᵢ, aᵢⱼ, sᵢⱼₖₗ, hₖ, tₗ)

### Components:
- **eᵢ**: Entity (e.g., *Galaxy ZX Pro*)
- **aᵢⱼ**: Aspect of the entity (e.g., *display*, *battery*)
- **sᵢⱼₖₗ**: Sentiment expressed (e.g., *positive*, *negative*, *neutral*, or rating like *4 stars*)
- **hₖ**: Opinion holder (e.g., *Sarah*, *Sarah’s brother*)
- **tₗ**: Time when the opinion was expressed (e.g., *March 5, 2022*)

---

### Index Explanation Table

| **Index** | **Meaning**                                      | **Example**                                  |
|-----------|--------------------------------------------------|----------------------------------------------|
| i         | Identifies the entity                           | e₁ = Galaxy ZX Pro; e₂ = iPhone 15          |
| j         | Identifies the aspect of entity i              | a₁₁ = display; a₁₂ = battery; a₁₃ = camera  |
| k         | Identifies the opinion holder                  | h₁ = Sarah; h₂ = Sarah’s brother            |
| l         | Identifies the time of sentiment               | t₁ = March 5, 2022; t₂ = April 10, 2022     |
| sᵢⱼₖₗ    | Sentiment about aspect j of entity i by holder k at time l | positive, negative, neutral, or rating |


In [0]:
pip install nltk

Note: you may need to restart the kernel using %restart_python or dbutils.library.restartPython() to use updated packages.


# Practice: Movie Review Classification
In this application activity, we will explore **sentiment data** and apply **classification techniques** using the `movie_reviews` dataset from **NLTK**—one of the most widely used natural language toolkits in Python.

## Objective
- Understand how to preprocess and analyze sentiment data.
- Apply machine learning models for **binary sentiment classification** (positive vs. negative).
- Evaluate performance using metrics like **accuracy**, **precision**, **recall**, and **F1-score**.

### Key Steps:
1. **Load the dataset** using NLTK.
2. **Explore the data**.
3. **Establish the Pipeline**:
4. **Model training and prediction** 
5. **Evaluate performance** using classification metrics.


In [0]:
#imports and data loading
import pandas as pd
import os
import nltk
nltk.download('movie_reviews')
from nltk.corpus import movie_reviews

[nltk_data] Downloading package movie_reviews to
[nltk_data]     /home/spark-41d57dc5-a04f-4db4-a2a8-0b/nltk_data...
[nltk_data]   Package movie_reviews is already up-to-date!


**If you are curious to see the whole dataset, you can download all reviews of this dataset as follows:**

In [0]:
# Create main folder
base_dir = 'movie_review'
os.makedirs(base_dir, exist_ok=True)

# Save reviews to appropriate subfolders
for fileid in movie_reviews.fileids(): # iterating over all reviews in the NLTK movie_reviews.
    category = movie_reviews.categories(fileid)[0]  # extracting 'pos' or 'neg'
    review_text = movie_reviews.raw(fileid) # retrieving the full text of the review as a string

    # Create subfolder for category
    category_dir = os.path.join(base_dir, category) # building the folder path using base_dir
    os.makedirs(category_dir, exist_ok=True)

    # Save review using fileid
    filename = os.path.basename(fileid)
    filepath = os.path.join(category_dir, filename)

    # Write review text
    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(review_text)

#You may want to save the reviews in a structured dataframe for further processing. 
docs = [(movie_reviews.raw(fileid), movie_reviews.categories(fileid)[0])
        for fileid in movie_reviews.fileids()]

moveis_review_df = pd.DataFrame(docs, columns=['review', 'label'])
print(moveis_review_df.head())


                                              review label
0  plot : two teen couples go to a church party ,...   neg
1  the happy bastard's quick movie review \ndamn ...   neg
2  it is movies like these that make a jaded movi...   neg
3   " quest for camelot " is warner bros . ' firs...   neg
4  synopsis : a mentally unstable man undergoing ...   neg


**Exploring the categories and names of the review files**

In [0]:
print(movie_reviews.categories()) # Show the categories
print(movie_reviews.fileids()[:10]) # Show the first 10 review file IDs

['neg', 'pos']
['neg/cv000_29416.txt', 'neg/cv001_19502.txt', 'neg/cv002_17424.txt', 'neg/cv003_12683.txt', 'neg/cv004_12641.txt', 'neg/cv005_29357.txt', 'neg/cv006_17022.txt', 'neg/cv007_4992.txt', 'neg/cv008_29326.txt', 'neg/cv009_29417.txt']


**Exploring one review.**

In [0]:
fileid = movie_reviews.fileids()[0] # get the first file ID from the dataset.
print("File ID:", fileid)
print("Category:", movie_reviews.categories(fileid)[0]) # print the label pos or neg
print("Review text:\n", movie_reviews.raw(fileid)) # print the content of the review

File ID: neg/cv000_29416.txt
Category: neg
Review text:
 plot : two teen couples go to a church party , drink and then drive . 
they get into an accident . 
one of the guys dies , but his girlfriend continues to see him in her life , and has nightmares . 
what's the deal ? 
watch the movie and " sorta " find out . . . 
critique : a mind-fuck movie for the teen generation that touches on a very cool idea , but presents it in a very bad package . 
which is what makes this review an even harder one to write , since i generally applaud films which attempt to break the mold , mess with your head and such ( lost highway & memento ) , but there are good and bad ways of making all types of films , and these folks just didn't snag this one correctly . 
they seem to have taken this pretty neat concept , but executed it terribly . 
so what are the problems with the movie ? 
well , its main problem is that it's simply too jumbled . 
it starts off " normal " but then downshifts into this " fantasy 

**Find the average review length.**

In [0]:
lengths = [len(movie_reviews.words(fileid)) for fileid in movie_reviews.fileids()]
print(f"Average review length (words): {sum(lengths) / len(lengths):.2f}")

Average review length (words): 791.91


**To get class distributions, how many pos and neg reviews in the dataset.
**

In [0]:
from collections import Counter
labels = [movie_reviews.categories(fileid)[0] for fileid in movie_reviews.fileids()]
print(Counter(labels))

Counter({'neg': 1000, 'pos': 1000})


**Classification**

In [0]:
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import LinearSVC
from sklearn.metrics import classification_report
import random

**Preparing the documents.** <br>
In the following code snippet, the function movie_reviews.categories() returns the two sentiment labels available in the dataset: 'pos' for positive and 'neg' for negative. 
For each category, movie_reviews.fileids(category) retrieves the list of file IDs (unique identifiers for each review) that belong to that sentiment. Then, movie_reviews.raw(fileid) is used to load the full text content of a given review. 
By iterating through each file ID within both sentiment categories, you construct the variable _documents_ which is a list of tuples each containing the raw review text and its associated sentiment label. 

In [0]:
documents = [(movie_reviews.raw(fileid), category)
             for category in movie_reviews.categories()
             for fileid in movie_reviews.fileids(category)]


**Shuffle the data to avoid any bias due to ordering.**

In [0]:
random.shuffle(documents)

**Separate the data into features (text) and labels. Then Split into training and test sets.**

In [0]:
texts = [text for text, label in documents] # list of review texts
labels = [label for text, label in documents] # list of corresponding labels ('pos' or 'neg')

X_train, X_test, y_train, y_test = train_test_split(
    texts, labels, test_size=0.2, random_state=42)


**Create a pipeline with TF-IDF and SVM.** <br>
Note that <br>
•	('tfidf', TfidfVectorizer(stop_words='english', ngram_range=(1, 2))): converts the reviews into numerical features using the TF-IDF method. <br>
•	 stop_words='english' removes common English words like “the”, “is”, “and” that carry little sentiment. <br>
•	ngram_range=(1, 2) tells it to include: Unigrams (e.g., "good", "excellent") and Bigrams (e.g., "very good", "not bad") <br>
•	'svm', LinearSVC()): applies Support Vector Machine (SVM) classifier with a linear kernel


In [0]:
pipeline = Pipeline([
    ('tfidf', TfidfVectorizer(stop_words='english', ngram_range=(1, 2))),
    ('svm', LinearSVC())
])

**Train the classifier.**

In [0]:
pipeline.fit(X_train, y_train)

Pipeline(steps=[('tfidf',
                 TfidfVectorizer(ngram_range=(1, 2), stop_words='english')),
                ('svm', LinearSVC())])

**Predict and evaluate.**

In [0]:
y_pred = pipeline.predict(X_test)
print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

         neg       0.83      0.77      0.80       197
         pos       0.79      0.84      0.82       203

    accuracy                           0.81       400
   macro avg       0.81      0.81      0.81       400
weighted avg       0.81      0.81      0.81       400



**Interpretaion** <br>


### What does this mean? 

#### Negative Reviews
- **Precision**: 83% of predicted negatives were correct.
- **Recall**: 80% of actual negatives were correctly identified.
- **F1-Score**: 0.81 — harmonic mean of precision and recall.
- **Support**: 188 negative reviews in the test set.

#### Positive Reviews
- **Precision**: 83% of predicted positives were correct.
- **Recall**: 85% of actual positives were correctly identified.
- **F1-Score**: 0.84 — balance between precision and recall.
- **Support**: 212 positive reviews in the test set.

#### Overall Accuracy
- **Accuracy**: 83% of reviews were correctly classified.
- **Support**: 400 reviews in the test set.


#### Average Scores
- **Macro Average**: 0.83  
  - Unweighted average across all classes.  
  - Example: Macro F1 = (F1_negative + F1_positive) / 2  
  - Treats all classes equally, may be misleading on imbalanced datasets.

- **Weighted Average**: 0.83  
  - Weighted by the number of instances in each class.  
  - Example: Weighted F1 = (F1_neg × Support_neg + F1_pos × Support_pos) / Total_support  
  - Reflects overall classifier performance, accounting for class imbalance.

> Note that: The macro and weighted averages are equal here because the dataset is balanced (400 reviews total).
